src/main/java/org/openstreetmap/josm/plugins/pt_assistant/actions/ExtractRelationMembersToNewRelationAction.java

```diff
@@ -0,0 +1,491 @@
+                        if (cbProposed.isSelected()) {
+                            extractedRelation.put("state", "proposed");
+                            extractedRelation.put("name", tfNameTag.getText() + " (wenslijn)");
+                        }
+                        if (cbDeviation.isSelected()) {
+                            extractedRelation.put("name", tfNameTag.getText() + " (omleiding)");
+                        }
+                        DataSet activeDataSet = getLayerManager().getActiveDataSet();
+                        commands.add(new AddCommand(activeDataSet, extractedRelation));
+                    }
+                    commands.add(new ChangeCommand(originalRelation, clonedRelation));
+                    addExtractedRelationToParentSuperrouteRelations(originalRelation, commands, parentRelations, selectedIndices, extractedRelation);
+                    UndoRedoHandler.getInstance().add(new SequenceCommand(I18n.tr("Extract ways to relation"), commands));
+                    RelationEditor extraEditor = RelationEditor.getEditor(getEditor().getLayer(), extractedRelation, emptyList());
+                    extraEditor.setVisible(true);
+                    extraEditor.setAlwaysOnTop(true);
+
+                    try {
+                        editor.reloadDataFromRelation();
+                        // todo this often gives index out of bounds exception
+                        // I can't figure out why.
+                        // catching it here causes the exception to occur when
+                        // the user presses the update button, so not really a solution
+                        // The update button appears if I remove 1 or 2 members before extracting
+                        // so when the relation editor is 'dirty'.
+                        // using memberTableModel.applyToRelation(originalRelation);
+                        // at the beginning doesn't seem to actually save the relation
+                    } catch (Exception e) {
+                        Logging.error(e);
+                    }
+                }
+            } else {
+                splitInSegments(originalRelation);
+                editor.reloadDataFromRelation();
+            }
+        }
+    }
+
+    public void splitInSegments(Relation originalRelation) {
+        final DataSet activeDataSet = getLayerManager().getActiveDataSet();
+        final Relation clonedRelation = new Relation(originalRelation);
+        final long clonedRelationId = clonedRelation.getId();
+        final List<RelationMember> members = clonedRelation.getMembers();
+        List<PTSegmentToExtract> segments = new ArrayList<>();
+        RelationMember previousWayMember = null;
+        RelationMember currentWayMember = null;
+        RelationMember nextWayMember = null;
+        boolean startNewSegment = false;
+        PTSegmentToExtract segment = null;
+        for (int i = members.size()-1; i>=0; i--) {
```

Comment:
If you are only interested in public transport ways, I'd filter the list before looping:
```java
final List<Way> ptWays = clonedRelation.getMembers().stream().filter(RelationMember::isWay).filter(RouteUtils::isPTWay).map(RelationMember::getWay).collect(Collectors.toList());
```

Then you can do this:
```java
for (int i = members.size() - 2; i >= 1; i--) {
  final Way previousWay = ptWays.get(i - 1);
  final Way currentWay = ptWays.get(i);
  final Way nextWay = ptWays.get(i + 1);
  …
}
```
so you don't have to shuffle values around between these three variables.

Reply:
I tried to do this, but I need to decrement the counter for each relation member encountered, as I'm using that counter to compile a list of indexes, to remove those members from the cloned relation.

I'm going to try a little bit more, but at the moment it's completely broken (again). So it will take a lot of single stepping to figure out what is going on.

Is there a way of keeping track of the indexes of those Way members when using the .filter() expression? i.e. can it be a list of Pairs?